packages/mux-uploader/src/progress.ts

```diff
@@ -0,0 +1,167 @@
+import { globalThis, document } from 'shared-polyfills';
+
+const template = document.createElement('template');
+const TYPES = {
+  BAR: 'bar',
+  RADIAL: 'radial',
+};
+
+const ariaDescription = 'Media upload progress bar';
+
+template.innerHTML = `
+<style>
+  :host {
+    position: relative;
+  }
+
+  .bar-type {
+    background: #e6e6e6;
+    border-radius: 100px;
+    position: relative;
+    height: 4px;
+    width: 100%;
+  }
+  
+  .radial-type, .bar-type {
+    display: none;
+  }
+
+  :host([type="radial"][upload-in-progress]) .radial-type {
+    display: block;
+  }
+  
+  :host([type="bar"][upload-in-progress]) .bar-type {
+    display: block;
+  }
+
+  :host([type="bar"][upload-error]) .progress-bar {
+    background: #e22c3e;
+  }
+
+  .progress-bar {
+    box-shadow: 0 10px 40px -10px #fff;
+    border-radius: 100px;
+    background: var(--progress-bar-fill-color, #000000);
+    height: 4px;
+    width: 0%;
+  }
+
+  circle {
+    stroke: var(--progress-radial-fill-color, black);
+    stroke-width: 6;  /* Thickness of the circle */
+    fill: transparent; /* Make inside of the circle see-through */
+  
+    /* Animation */ 
+    transition: 0.35s;
+    transform: rotate(-90deg);
+    transform-origin: 50% 50%;
+    -webkit-transform-origin: 50% 50%;
+    -moz-transform-origin: 50% 50%;
+  }
+</style>
+
+<slot></slot>
+<div class="bar-type">
+  <div role="progressbar" aria-valuemin="0" aria-valuemax="100" class="progress-bar" id="progress-bar" tabindex="0"></div>
+</div>
+<div class="radial-type">
+  <svg
+    width="120"
+    height="120">
+    <!-- To prevent overflow of the SVG wrapper, radius must be  (svgWidth / 2) - (circleStrokeWidth * 2)
+      or use overflow: visible on the svg.-->
+    <circle
+      r="52"
+      cx="60"
+      cy="60"
+    />
+  <svg>
+</div>
+`;
+
+class Progress extends globalThis.HTMLElement {
+  svgCircle: SVGCircleElement | null | undefined;
+  progressBar: HTMLElement | null | undefined;
+
+  $isUploading = false;
+
+  constructor() {
+    super();
+    const shadowRoot = this.attachShadow({ mode: 'open' });
+    shadowRoot.appendChild(template.content.cloneNode(true));
+
+    this.svgCircle = this.shadowRoot?.querySelector('circle');
+    this.progressBar = this.shadowRoot?.getElementById('progress-bar');
+    this.progressBar?.setAttribute('aria-description', ariaDescription);
+  }
+
+  connectedCallback() {
+    this.setDefaultType();
+    const parent = document.querySelector('mux-uploader');
+
+    parent?.addEventListener('progress', (e) => {
```

Comment:
suggestion(non-blocking): We as a followup, we should probably make the uploader "push" progress updates to this component, rather than the current impl that "reaches out" for progress updates. This avoids unnecessary assumptions in the DOM structure, avoids potential memory leaks/"dangling references", and allows the component to be more "unit testable".